src/components/tl/mlx5/mcast/tl_mlx5_mcast_allgather.c

```diff
@@ -0,0 +1,439 @@
+    }
+
+    return status;
+}
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_prepare_allgather(void* sbuf, void *rbuf, int size,
+                                                               ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                               ucc_tl_mlx5_mcast_coll_req_t *req)
+{
+    ucc_tl_mlx5_mcast_reg_t *reg = NULL;
+    ucc_status_t             status;
+
+    req->comm   = comm;
+    req->ptr    = sbuf;
+    req->rptr   = rbuf;
+    req->length = size;
+    req->mr     = comm->pp_mr;
+    req->rreg   = NULL;
+    /* - zero copy protocol only provides zero copy design at sender side
+     * - truly zero copy protocol provides zero copy design at receiver side as well
+     * here we select the sender side protocol */
+    req->proto  = (req->length < comm->max_eager) ? MCAST_PROTO_EAGER :
+                                                    MCAST_PROTO_ZCOPY;
+
+    if (comm->commsize > ONE_SIDED_RELIABILITY_MAX_TEAM_SIZE) {
+        tl_warn(comm->lib,
+                "team size is %d but max supported team size of mcast allgather is %d",
+                comm->commsize, ONE_SIDED_RELIABILITY_MAX_TEAM_SIZE);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    req->offset      = 0;
+    req->num_packets = (req->length + comm->max_per_packet - 1)/comm->max_per_packet;
+
+    if (req->num_packets == 0) {
+        req->num_packets = 1;
+    }
+
+    ONE_SIDED_MAX_PACKET_COUNT(comm->allgather_comm.max_num_packets);
+
+    if (comm->allgather_comm.max_num_packets < req->num_packets) {
+        tl_warn(comm->lib,
+                "msg size is %ld but max supported msg size of mcast allgather is %d",
+                req->length, comm->allgather_comm.max_num_packets * comm->max_per_packet);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    req->last_pkt_len = req->length - (req->num_packets - 1)*comm->max_per_packet;
+
+    ucc_assert(req->last_pkt_len > 0 && req->last_pkt_len <= comm->max_per_packet);
+
+    if (req->proto == MCAST_PROTO_ZCOPY) {
+        /* register the send buffer */
+       status = ucc_tl_mlx5_mcast_mem_register(comm->ctx, req->ptr, req->length, &reg);
+       if (UCC_OK != status) {
+            return status;
+       }
+       req->rreg = reg;
+       req->mr   = reg->mr;
+    }
+
+    if (comm->one_sided.reliability_enabled) {
+        req->one_sided_reliability_scheme = (req->length <
+                comm->one_sided.reliability_scheme_msg_threshold) ?
+                ONE_SIDED_ASYNCHRONOUS_PROTO : ONE_SIDED_SYNCHRONOUS_PROTO;
+    } else {
+        req->one_sided_reliability_scheme = ONE_SIDED_NO_RELIABILITY;
+    }
+
+    req->ag_counter = comm->allgather_comm.coll_counter;
+    req->to_send    = req->num_packets;
+    req->to_recv    = comm->commsize * req->num_packets;
+
+    comm->allgather_comm.coll_counter++;
+    return UCC_OK;
+}
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_coll_do_allgather(void* sbuf, void *rbuf, int size,
```

Comment:
There are two different functions with names `ucc_tl_mlx5_mcast_coll_do_allgather` and `ucc_tl_mlx5_mcast_do_allgather`. The functions perform very different things and their name do not indicate what each function is doing. For example, here, allgather is not posted nor progressed, only some setup are handled, so "do_allgather" is misleading.

Reply:
Thanks for noticing this. I am fixing it in the commit. Since we have two different algorithms for Allgather, let me name one of the staging-based (which is in this PR) and zero-copy (which I removed from this PR and will be a separate PR).